compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp

```diff
@@ -0,0 +1,227 @@
+    for (int i = 0; i < static_cast<int32_t>(size); ++i)
+      node->at<loco::DataType::U8>(i) = 0;
+
+    *scaling_factor = 1;
+    *zp = 0;
+    return;
+  }
+
+  compute_asym_scale_zp(min, max, scaling_factor, zp);
+  const float scaling_factor_inv = 1.0 / *scaling_factor;
+  std::vector<int32_t> quantized_values(size);
+  for (int i = 0; i < static_cast<int32_t>(size); ++i)
+  {
+    quantized_values[i] = static_cast<int32_t>(
+        std::round(*zp + node->at<loco::DataType::FLOAT32>(i) * scaling_factor_inv));
+  }
+
+  node->dtype(loco::DataType::U8);      // change the type of tensor
+  node->size<loco::DataType::U8>(size); // resize tensor
+  for (int i = 0; i < static_cast<int32_t>(size); ++i)
+  {
+    node->at<loco::DataType::U8>(i) = std::min(kMaxScale, std::max(kMinScale, quantized_values[i]));
+  }
+}
+
+bool is_quantized(const CircleNode *node)
+{
+  return node->dtype() == loco::DataType::U8 || // activation, weight
+         node->dtype() == loco::DataType::S32;  // bias
+}
+
+// Check if node is weights of conv2d, depthwise_conv2d, or fully_connected layer
+bool is_weights(CircleNode *node)
+{
+  auto circle_const = dynamic_cast<CircleConst *>(node);
+  if (circle_const == nullptr)
+    return false;
+
+  auto succs = loco::succs(node);
+  if (succs.size() != 1) // assume weights is used by only one node
+    return false;
+
+  for (auto out : succs)
+  {
+    auto conv = dynamic_cast<CircleConv2D *>(out);
+    if (conv != nullptr && conv->filter() == circle_const)
+      return true;
+
+    auto dw_conv = dynamic_cast<CircleDepthwiseConv2D *>(out);
+    if (dw_conv != nullptr && dw_conv->filter() == circle_const)
+      return true;
+
+    auto fc = dynamic_cast<CircleFullyConnected *>(out);
+    if (fc != nullptr && fc->weights() == circle_const)
+      return true;
+  }
+  return false;
+}
+
+/**
+ * @brief QuantizeDequantizeWeights quantizes and dequantizes tensors for weights
+ * @details Find min/max values on the fly, quantize the model, and dequantize the model
+ */
+struct QuantizeDequantizeWeights final : public luci::CircleNodeMutableVisitor<bool>
+{
+  QuantizeDequantizeWeights(loco::DataType input, loco::DataType output)
+      : input_type(input), output_type(output)
+  {
+  }
+
+  loco::DataType input_type;
+  loco::DataType output_type;
+
+  // Quantize and dequantize input tensors of each node
+  bool visit(luci::CircleNode *node)
+  {
+    LOGGER(l);
+    INFO(l) << "QuantizeDequantizeWeights visit node: " << node->name() << std::endl;
+    auto arity = node->arity();
+    for (uint32_t i = 0; i < arity; i++)
+    {
+      auto input_node = node->arg(i);
+      auto circle_node = loco::must_cast<luci::CircleNode *>(input_node);
+
+      // Check if this is already quantized
+      if (is_quantized(circle_node))
+        continue;
+
+      if (is_weights(circle_node))
+      {
+        auto circle_const = dynamic_cast<luci::CircleConst *>(circle_node);
```

Comment:
Can you please use `loco::must_cast()` ? Just to prevent static analysis tool warnings (but not sure)...

Reply:
Changed. Thanks.